### ISSUE TITLE
`vue/component-name-in-template-casing` is not ignoring some built-in components

### ISSUE THREAD

**Issue by valentinoli** [vuejs/eslint-plugin-vue#1722]
**Checklist**

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have read the [FAQ](https://eslint.vuejs.org/user-guide/#faq) and my problem is not listed.

**Tell us about your environment**

- **ESLint version:** 7.32.0
- **eslint-plugin-vue version:** 7.17.0
- **Node version:** 14.17.6
- **Operating System:** Windows 10

**Please show your full configuration:**

```json5
module.exports = {
  root: true,
  env: {
    browser: true,
    node: true
  },
  parserOptions: {
    parser: '@babel/eslint-parser',
    requireConfigFile: false
  },
  extends: [
    '@nuxtjs',
    'plugin:prettier/recommended',
    'plugin:nuxt/recommended'
  ],
  plugins: ['prettier'],
  // add your custom rules here
  rules: {
    'no-console': [
      'warn',
      {
        allow: ['info', 'warn', 'error']
      }
    ],
    'vue/valid-v-slot': [
      'error',
      {
        allowModifiers: true
      }
    ],
    'vue/no-unregistered-components': ['off', {}],
    'vue/component-name-in-template-casing': [
      'error',
      'PascalCase',
      {
        registeredComponentsOnly: false
      }
    ]
  }
}
```

**What did you do?**	

```vue
<template>
  <component :is="someComponent"></component>
</template>
```

**What did you expect to happen?**
As the [docs explain](https://eslint.vuejs.org/rules/no-unregistered-components.html) in the note, I expected the `vue/component-name-in-template-casing` rule to ignore the built-in `<component>`

**What actually happened?**

Raw output from ESLint
```
Component name "component" is not PascalCase. eslint(vue/component-name-in-template-casing)
```

Note: Same story for `<client-only>` and `<keep-alive>` built-ins. However, `<slot>` and `<template>` are ignored.

**Repository to reproduce this issue**

https://github.com/hestiaAI/hestialabs-experiences/

I added the following configuration to `.eslintrc.js` to fix the issue temporarily:

```js
ignores: ['/^component|client-only|keep-alive$/']
```

**Comment by FloEdelmann** [comment]
Can't you use `<Component>`, `<ClientOnly>` and `<KeepAlive>`?

However, the docs clearly don't match the behavior, so either of them needs to be adjusted.

**Comment by valentinoli** [comment]
I think that would be inconsistent with our preference of having built-in components kebab-cased

**Comment by valentinoli** [comment]
I noticed this file, but it is not imported anywhere ... except in the vue3 builtin file which is in turn not imported anywhere 

https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/utils/vue2-builtin-components.js

**Comment by FloEdelmann** [comment]
> our preference of having built-in components kebab-cased

I think this rule is the only one where we enforce the casing of built-in components, isn't it? In the Vue docs, they are kebab-case, but so are custom components: https://v3.vuejs.org/guide/component-basics.html

However, in the relevant style guide section, built-in components are not explicitly mentioned: https://v3.vuejs.org/style-guide/#component-name-casing-in-templates-strongly-recommended

So I think it's a matter of taste. IMO, they should not be excluded, but the docs should be adjusted to explicitly mention that they are included. Users who want to exclude them can use the `ignores` option like you did.

> this file, but it is not imported anywhere

I don't know whether/how this file is/was used. @ota-meshi Can you help here?

**Comment by ota-meshi** [comment]
> So I think it's a matter of taste.

I agree with you.
If we change the rule, I think it should be an option to ignore the built-in components.

> I don't know whether/how this file is/was used. @ota-meshi Can you help here?

Used it here:
https://github.com/vuejs/eslint-plugin-vue/blob/master/lib/utils/vue3-builtin-components.js

**Comment by FloEdelmann** [comment]
@valentinoli I just noticed that the [note](https://eslint.vuejs.org/rules/no-unregistered-components.html) you linked to is in the docs of another rule: `vue/no-unregistered-components`, not `vue/component-name-in-template-casing`.

So I think there is nothing that needs to be done here.

* Built-in components are not reported if `registeredComponentsOnly` is `true` (default).
* They are reported if `registeredComponentsOnly` is set to `false`.
* They can be excluded with the `ignores` option.
* So they are not handled differently than custom components. The docs don't imply anything different.

Since the behavior is expected, I added test cases to document this in #1737.

**Comment by FloEdelmann** [comment]
@valentinoli Note that you should adapt your `ignores` option:

Your regex `/^component|client-only|keep-alive$/` matches everything that starts with `component`, anything that contains `client-only` and anything that ends with `keep-alive`.

Either add parentheses to the regex:

```diff
- ignores: ['/^component|client-only|keep-alive$/']
+ ignores: ['/^(component|client-only|keep-alive)$/']
```

Or (even better) use separate strings:

```diff
- ignores: ['/^component|client-only|keep-alive$/']
+ ignores: ['component', 'client-only', 'keep-alive']
```

**Comment by valentinoli** [comment]
> I just noticed that the note you linked to is in the docs of another rule: vue/no-unregistered-components, not vue/component-name-in-template-casing

My bad! Sorry about that.

**Comment by valentinoli** [comment]
> Note that you should adapt your ignores option

Much appreciated tip, thank you!

**Comment by valentinoli** [comment]
> Used it here:\nhttps://github.com/vuejs/eslint-plugin-vue/blob/master/lib/utils/vue3-builtin-components.js

Yes, but where is that file used?

**Comment by ota-meshi** [comment]
You can use the search.
https://github.com/vuejs/eslint-plugin-vue/search?q=vue3-builtin-components

**Comment by valentinoli** [comment]
Ok, I thought I already did that. Sorry and thanks!
